### PR TITLE
feat: Correctness of Memcpy program

### PR DESCRIPTION
**Pull request by bollu**
We add a manual proof of `memcpy` to understand just how complex a proof will be. The proof challenges `simp_mem`, due to, what we suspect, are `omega` scaling bottlenecks. This file will be used as a litmus test for improving automation.

### Testing:

`make all` succeeds, nothing was change with regards to conformance.

### License:

By submitting this pull request, I confirm that my contribution is
made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Proofs/Experiments/Memcpy/MemCpyVCG.lean [hunk 1]
```diff
@@ -0,0 +1,831 @@
...
+import Arm
+import Tactics.CSE
+import Tactics.Rename
+import Tactics.Sym
+import Tactics.Rename
+import Tactics.Name
+import Tactics.ClearNamed
+import Tactics.StepThms
+import Correctness.ArmSpec
+import Arm.Insts.Common
```
Comment (shigoel) [inline Proofs/Experiments/Memcpy/MemCpyVCG.lean:18]:
```suggestion
```
Isn't this imported by `Arm`?

File: Proofs/Experiments/Memcpy/MemCpyVCG.lean [hunk 2]
```diff
@@ -0,0 +1,831 @@
...
+  However, we prefer using quantifiers because it mirrors how we tend to state
+  and prove some cryptographic obligations, i.e., each `n`-byte block in a given
+  region is hashed/encrypted correctly (for some appropriate `n`, which is 16
+  here).
+  -/
+  (∀ i : BitVec 64, i < num_blks →
+    read_mem_bytes 16 (dst_base + (16 * i)) sf =
+    id (read_mem_bytes 16 (src_base + (16 * i)) s0)) ∧
+  -- TODO (@bollu): we can't prove this, because we don't have this in the loop invariant
+  -- to show that emory regions separate from the destination are unchanged.
```
Comment (shigoel) [inline Proofs/Experiments/Memcpy/MemCpyVCG.lean:148]: Old TODO?

File: Proofs/Experiments/Memcpy/MemCpyVCG.lean [hunk 3]
```diff
@@ -0,0 +1,831 @@
...
+def cut (s : ArmState) : Bool :=
+  -- First instruction
+  r StateField.PC s = 0x8e0#64 ||
+  -- Loop guard (branch instruction)
+  r StateField.PC s = 0x8f4#64 ||
+  -- First instruction following the loop
+  -- which also happens to be the program's last instruction
+  r StateField.PC s = 0x8f8#64
+
+structure LoopInv (s0 si : ArmState)
```
Comment (shigoel) [inline Proofs/Experiments/Memcpy/MemCpyVCG.lean:172]: Where's this used?

**Comment by shigoel** [comment]
OK with this PR, modulo missing PR message and nits, esp. since this proof will be redone using an integrated `sym_n` + `simp_mem`.

**Comment by bollu** [comment]
@shigoel comments addressed!

File: Proofs/Experiments/Memcpy/MemCpyVCG.lean [hunk 4]
```diff
@@ -15,11 +15,13 @@ import Tactics.Name
 import Tactics.ClearNamed
 import Tactics.StepThms
 import Correctness.ArmSpec
-import Arm.Insts.Common
 import Arm.Memory.SeparateAutomation
 import Arm.Syntax
 import Tactics.SkipProof
 
+-- Disable linters, they take too much time.
```
Comment (shigoel) [inline Proofs/Experiments/Memcpy/MemCpyVCG.lean:22]: Sigh. Is it the unused vars check again, or something else? I believe the former's rehaul is in progress.

Comment (bollu) [reply Proofs/Experiments/Memcpy/MemCpyVCG.lean:22]: Yes, the same!~

Comment (shigoel) [reply Proofs/Experiments/Memcpy/MemCpyVCG.lean:22]: Well then why don't we do just the following:
`set_option linter.unusedVariables false in...`

Comment (bollu) [reply Proofs/Experiments/Memcpy/MemCpyVCG.lean:22]: Thanks for forcing me to be precise :) Done.

**Review by shigoel — APPROVED** [review]
Thanks!

**Pull request merged**